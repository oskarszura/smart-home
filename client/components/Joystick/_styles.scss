$c-joystick-size: 100px;
$c-joystick-knob-color: #bebfc1;
$c-joystick-color: #f2f2f2;

.c-joystick {
  position: relative;
  background-color: $c-joystick-color;
  width: $c-joystick-size;
  height: $c-joystick-size;
  border-radius: $c-joystick-size * 0.5;
  cursor: pointer;

  &__knob {
    position: absolute;
    background-color: $c-joystick-knob-color;
    width: $c-joystick-size * 0.5;
    height: $c-joystick-size * 0.5;
    border-radius: $c-joystick-size * 0.25;

    .c-joystick--inactive & {
      opacity: 0.2;
    }
  }

  &__drag {
    position: absolute;
    opacity: 0;
    width: $c-joystick-size * 0.5;
    height: $c-joystick-size * 0.5;
    border-radius: $c-joystick-size * 0.25;
  }
}
