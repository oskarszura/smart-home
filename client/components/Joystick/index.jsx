// @flow
import _ from 'lodash';
import React from 'react';
import ReactDOM from 'react-dom';

type Props = {
  onPositionChange: (left: number, top: number) => void,
  isEnabled: boolean,
};

type State = {
  left: number,
  top: number,
};

const JOYSTICK_RADIUS = 50;

class Joystick extends React.PureComponent<Props, State> {
  static onDragOver(event: SyntheticDragEvent<HTMLDivElement>) {
    event.preventDefault();
  }

  static defaultProps = {
    isEnabled: true,
  };

  constructor() {
    super();
    this.state = {
      left: JOYSTICK_RADIUS * 0.5,
      top: JOYSTICK_RADIUS * 0.5,
    };
  }

  onMove(event: SyntheticDragEvent<Element>) {
    event.preventDefault();
    const { clientX, clientY } = event;

    /* eslint-disable react/no-find-dom-node */
    const node: null | Element | Text = ReactDOM.findDOMNode(this);
    /* eslint-enable react/no-find-dom-node */
    if (node === null || node instanceof Text) {
      return;
    }
    const rect: ClientRect = node.getBoundingClientRect();
    const mappedX = clientX - rect.left - JOYSTICK_RADIUS * 0.5;
    const mappedY = clientY - rect.top - JOYSTICK_RADIUS * 0.5;

    this.move(mappedX, mappedY);
  }

  onDragEnd(event: SyntheticDragEvent<Element>) {
    event.preventDefault();
    this.move(JOYSTICK_RADIUS * 0.5, JOYSTICK_RADIUS * 0.5);
  }

  move(x: number, y: number) {
    const left = _.min([_.max([x, 0]), JOYSTICK_RADIUS]);
    const top = _.min([_.max([y, 0]), JOYSTICK_RADIUS]);

    const { onPositionChange } = this.props;

    onPositionChange(left, top);
    this.setState({
      left,
      top,
    });
  }

  render() {
    const { left, top } = this.state;
    const { isEnabled } = this.props;

    const classes = `c-joystick${isEnabled ? '' : ' c-joystick--inactive'}`;

    return (
      <div
        className={classes}
        onDragOver={event => Joystick.onDragOver(event)}
      >
        <div
          className="c-joystick__knob"
          style={{
            left,
            top,
          }}
        />
        <div
          className="c-joystick__drag"
          draggable={isEnabled}
          onDrag={(event: SyntheticDragEvent<HTMLDivElement>) => {
            this.onMove(event);
          }}
          onDragEnd={(event: SyntheticDragEvent<HTMLDivElement>) => {
            this.onDragEnd(event);
          }}
          style={{
            left,
            top,
          }}
        />
      </div>
    );
  }
}

export default Joystick;
