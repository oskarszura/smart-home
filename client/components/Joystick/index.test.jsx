import _ from 'lodash';
import React from 'react';
import { mount } from 'enzyme';
import Joystick from './index';

describe('Joystick', () => {
  it('should render correctly', () => {
    const component = mount(<Joystick onPositionChange={_.noop} />);

    expect(component).toMatchSnapshot();
  });
});
