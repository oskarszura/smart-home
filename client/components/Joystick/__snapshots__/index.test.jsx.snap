// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Joystick should render correctly 1`] = `
<Joystick
  isEnabled={true}
  onPositionChange={[Function]}
>
  <div
    className="c-joystick"
    onDragOver={[Function]}
  >
    <div
      className="c-joystick__knob"
      style={
        Object {
          "left": 25,
          "top": 25,
        }
      }
    />
    <div
      className="c-joystick__drag"
      draggable={true}
      onDrag={[Function]}
      onDragEnd={[Function]}
      style={
        Object {
          "left": 25,
          "top": 25,
        }
      }
    />
  </div>
</Joystick>
`;

exports[`Joystick should render correctly in disabled state 1`] = `
<Joystick
  isEnabled={false}
  onPositionChange={[Function]}
>
  <div
    className="c-joystick c-joystick--inactive"
    onDragOver={[Function]}
  >
    <div
      className="c-joystick__knob"
      style={
        Object {
          "left": 25,
          "top": 25,
        }
      }
    />
    <div
      className="c-joystick__drag"
      draggable={false}
      onDrag={[Function]}
      onDragEnd={[Function]}
      style={
        Object {
          "left": 25,
          "top": 25,
        }
      }
    />
  </div>
</Joystick>
`;
